packages/core-plugin-api/src/translation/TranslationRef.ts

```diff
@@ -34,6 +34,30 @@ export interface TranslationRef<
 /** @internal */
 type AnyMessages = { [key in string]: string };
 
+/** @ignore */
+type AnyNestedMessages = { [key in string]: AnyNestedMessages | string };
+
+/**
+ * Flattens a nested message declaration into a flat object with dot-separated keys.
+ *
+ * @ignore
+ */
+type FlattenedMessages<T extends AnyNestedMessages> = {
+  [K in keyof T]: (
+    x: T[K] extends infer V
+      ? V extends AnyNestedMessages
+        ? FlattenedMessages<V> extends infer FV
+          ? {
+              [P in keyof FV as `${K & string}.${P & string}`]: FV[P];
+            }
+          : never
+        : Pick<T, K>
+      : never,
+  ) => void;
+} extends Record<keyof T, (y: infer O) => void>
+  ? { [K in keyof O]: O[K] }
+  : never;
```

Comment:
![wizard](https://github.com/backstage/backstage/assets/8904624/af67d7d1-39c2-4092-9985-28306a6c24db)

Reply:
this guy types